File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/screens/pages/categories/CategoriesAdapter.java
package ru.usedesk.knowledgebase_gui.screens.pages.categories;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import ru.usedesk.common_gui.external.IUsedeskViewCustomizer;
import ru.usedesk.knowledgebase_gui.R;
import ru.usedesk.knowledgebase_sdk.external.entity.Category;


public class CategoriesAdapter extends RecyclerView.Adapter<CategoriesAdapter.SectionViewHolder> {

    private final List<Category> categoryList;
    private final IOnCategoryClickListener onCategoryClickListener;
    private final IUsedeskViewCustomizer usedeskViewCustomizer;

    CategoriesAdapter(@NonNull List<Category> categoryList,
                      @NonNull IOnCategoryClickListener onCategoryClickListener,
                      @NonNull IUsedeskViewCustomizer usedeskViewCustomizer) {
        this.categoryList = categoryList;
        this.onCategoryClickListener = onCategoryClickListener;
        this.usedeskViewCustomizer = usedeskViewCustomizer;
    }

    @NonNull
    @Override
    public SectionViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = usedeskViewCustomizer.createView(viewGroup, R.layout.usedesk_item_category);

        return new SectionViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull SectionViewHolder sectionViewHolder, int i) {
        sectionViewHolder.bind(categoryList.get(i), onCategoryClickListener);
    }

    @Override
    public int getItemCount() {
        return categoryList.size();
    }

    class SectionViewHolder extends RecyclerView.ViewHolder {

        private final TextView textViewTitle;

        SectionViewHolder(@NonNull View itemView) {
            super(itemView);

            textViewTitle = itemView.findViewById(R.id.tv_title);
        }

        void bind(@NonNull final Category category,
                  @NonNull final IOnCategoryClickListener onCategoryClickListener) {
            textViewTitle.setText(category.getTitle());

            itemView.setOnClickListener(v -> onCategoryClickListener.onCategoryClick(category.getId()));
        }
    }
}

File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/pages/categories/CategoriesAdapter.kt
package ru.usedesk.knowledgebase_gui.pages.categories

import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import ru.usedesk.common_gui.UsedeskBinding
import ru.usedesk.common_gui.inflateItem
import ru.usedesk.knowledgebase_gui.R
import ru.usedesk.knowledgebase_sdk.entity.UsedeskCategory

internal class CategoriesAdapter internal constructor(
        recyclerView: RecyclerView,
        lifecycleOwner: LifecycleOwner,
        private val viewModel: CategoriesViewModel,
        private val onCategoryClick: (Long) -> Unit
) : RecyclerView.Adapter<CategoriesAdapter.SectionViewHolder>() {

    private var categoryList = listOf<UsedeskCategory>()

    init {
        recyclerView.adapter = this

        viewModel.categoriesLiveData.observe(lifecycleOwner) {
            categoryList = it
            notifyDataSetChanged()
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): SectionViewHolder {
        return SectionViewHolder(inflateItem(viewGroup,
                R.layout.usedesk_item_category,
                R.style.Usedesk_KnowledgeBase) { rootView, defaultStyleId ->
            CategoryBinding(rootView, defaultStyleId)
        })
    }

    override fun onBindViewHolder(sectionViewHolder: SectionViewHolder, i: Int) {
        sectionViewHolder.bind(categoryList[i])
    }

    override fun getItemCount(): Int = categoryList.size

    inner class SectionViewHolder(
            private val binding: CategoryBinding
    ) : RecyclerView.ViewHolder(binding.rootView) {

        fun bind(category: UsedeskCategory) {
            binding.tvTitle.text = category.title
            binding.tvDescription.text = category.title//TODO: тут в сущности, видимо, должно быть поле с описанием
            binding.rootView.setOnClickListener {
                onCategoryClick(category.id)
            }
        }
    }

    internal class CategoryBinding(rootView: View, defaultStyleId: Int) : UsedeskBinding(rootView, defaultStyleId) {
        val tvTitle: TextView = rootView.findViewById(R.id.tv_title)
        val tvDescription: TextView = rootView.findViewById(R.id.tv_description)
        val tvCount: TextView = rootView.findViewById(R.id.tv_count)
    }
}